Enable "strict" compiler option

BigBahss opened the pull request:
Closes #61.

Changes have been made to adhere to strict type checking:
- Non-null assertions and definite-assignment assertions (`!` post-fix operator) have been added to please the compiler (these do not affect the compiled output). These assertions are not always strictly "safe", but are not any less safe than not having strict type checking at all. Plus, their presence in the code can be a signal that maybe the design should be changed to safely ensure that the types are not null/undefined.
- Some optional chaining (`?.`) has been been added when properties can be null/undefined.
- Some incorrect uses of `null` have been replaced with `undefined`.
- Add `defaultValue` arguments to `WorkspaceConfiguration.get()` so that it will not return undefined. The default values being passed are the default values defined in the package.json. Realistically, the function should not return undefined when default values are defined in the package.json, but this pleases the compiler.

The changes I've made were aimed at not changing the behavior of the extension, and weren't necessarily aimed at fixing any potential bugs. Strict type checking will make it less likely to introduce new bugs in future changes.

BigBahss commented on src/ast.ts:
```diff
@@ -39,11 +40,11 @@ class ASTFeature implements vscodelc.StaticFeature {
           // reveal(root) fails here: "Data tree node not found".
           if (adapter.hasRoot())
-          tree.reveal(null);
+            tree.reveal(null!);
```
This one is kind of silly since we are asserting that null is not null. But, the API expects a non-null object, and if passing null works around the issue described, then this should be fine.

sam-mccall replied:
I think maybe we should just @ts-ignore here instead, the current plan is for them to fix the typings: https://github.com/microsoft/vscode/commit/e95c40c1e461f83f5ff1c68d61253ab23c85f86b

BigBahss commented on src/semantic-highlighting.ts:
```diff
@@ -292,7 +292,7 @@ export class ThemeRuleMatcher {
   getBestThemeRule(scope: string): TokenColorRule {
     if (this.bestRuleCache.has(scope))
-      return this.bestRuleCache.get(scope);
+      return this.bestRuleCache.get(scope)!;
```
I would consider this a safe non-null assertion, since we have already checked that the `Map` has the value before getting it.

sam-mccall commented on src/ast.ts:
```diff
@@ -24,7 +24,8 @@ interface ASTNode {
 }
 const ASTRequestType =
-    new vscodelc.RequestType<ASTParams, ASTNode|null, void>('textDocument/ast');
+    new vscodelc.RequestType<ASTParams, ASTNode|undefined, void>(
```
I don't understand this one, the JSON-RPC response may literally be `null`

sam-mccall replied:
What I mean is, I think this type param corresponds to the actual JSON-RPC data sent by clangd, which definitely can be `null` (there's no `undefined` in JSON!).

So I think ASTRequestType needs to use `null` for correctness, but it's fine to convert that to `undefined` immediately after calling `sendRequest`.

BigBahss replied:
I was not aware of that. Then I agree, this should be changed back to `null` for the RequestType.

sam-mccall commented on src/ast.ts:
```diff
@@ -163,16 +164,16 @@ class TreeAdapter implements vscode.TreeDataProvider<ASTNode> {
   }
 
-  public getParent(node: ASTNode): ASTNode {
+  public getParent(node: ASTNode): ASTNode|undefined {
```
There are a bunch of places in this class where you've changed `null` to `undefined` - is there some fundamental reason why null is incorrect? (It makes sense to be consistent within a module of course.

My understanding is this is mostly a stylistic choice, and my preference would be to use null for "deliberately" null values, but this isn't my primary language...

sam-mccall replied:
No worries at all. I think if `undefined` is privileged by the `?.` operators etc then ergonomics is a good reason to prefer it.
(And it seems so, playground says `(null as any)?.foo` is `undefined`)
And this seems to be MS's preference, and it's mostly their APIs we're dealing with.

BigBahss replied:
I agree. So just to confirm, are deciding to use `undefined` for values stored by the extension and accepting `null` for incoming values (when necessary)?

sam-mccall replied:
That sounds good to me!

sam-mccall commented on src/clangd-context.ts:
```diff
@@ -58,7 +58,7 @@ export class ClangdContext implements vscode.Disposable {
       options: {cwd: vscode.workspace.rootPath || process.cwd()}
     };
-    const traceFile = config.get<string>('trace');
+    const traceFile = config.get<string>('trace', '');
```
I think that echoing defaults already given in the package.json is unacceptable duplication (and causes practical problems e.g. someone changing the default in the wrong place).

Can we add a `!` assertion in config.get() instead?

BigBahss replied:
That's fair. I believe the defaults can be retrieved through `WorkspaceConfiguration.inspect()`, but that may be more trouble than it's worth. Realistically adding asserts here would only cause a problem if a default value was removed from `package.json`, but I expect that's quite unlikely.

sam-mccall commented on src/config.ts:
```diff
@@ -17,7 +18,7 @@ export function getSecure<T>(key: string, workspaceState: vscode.Memento): T|
 // If there is workspace config, prompts the user and caches the decision.
 export async function getSecureOrPrompt<T>(
-    key: string, workspaceState: vscode.Memento): Promise<T> {
+    key: string, workspaceState: vscode.Memento): Promise<T|undefined> {
```
(again here I think we should assert the result rather than returning a maybe-result)

BigBahss replied:
I'll fix that.

sam-mccall commented on src/install.ts:
```diff
@@ -21,8 +21,9 @@ export async function activate(
       'clangd.update', async () => common.checkUpdates(true, ui)));
-  const status = await common.prepare(ui, config.get<boolean>('checkUpdates'));
-  return status.clangdPath;
+  const status =
```
The function comment says "returns the path or null", so I think we return status?.clangdPath instead maybe, and change the signature?

BigBahss replied:
Good catch, the caller of this function already checks if the result is falsy. I'll fix that and add `null` to the return type.

sam-mccall commented on src/memory-usage.ts:
```diff
@@ -80,8 +80,8 @@ class MemoryUsageFeature implements vscodelc.StaticFeature {
 
-  get root(): InternalTree|null { return this.root_; }
-  set root(n: InternalTree|null) {
+  get root(): InternalTree|undefined { return this.root_; }
```
again why these null->undefined changes?

BigBahss replied:
Addressed in a previous comment, but it's because `private root_?: InternalTree` is already declared as being `InternalTree | undefined`.

sam-mccall commented on tsconfig.json:
```diff
@@ -16,7 +16,7 @@
         "sourceMap": true,
         "rootDir": ".",
-        "alwaysStrict": true,
+        "strict": true,
```
It's surprising that "strict" is more strict than "alwaysStrict" :-)

BigBahss replied:
Yeah, it really seems like a "gcc -Wall" scenario haha.

sam-mccall reviewed (commented):
Thanks for doing this cleanup! And sorry if some of these comments are dumb, I'm not deeply familiar with TS's type system. I do think if we're going to enable strict mode then I should understand it though...

BigBahss reviewed (commented):
Addressed comments (further consensus is needed on the `null` vs `undefined` issue).

BigBahss commented:
I was going crazy there for a minute wondering how the CI was failing on a non-existent file when the tests passed on my machine 😅. Anyways I merged and modified the new file for strict type checking.

I fixed all the things that were mentioned. Additionally I changed all `==` and `!=` to `===` and `!==` since that's very closely related to strict type checking. Let me know if I missed anything or whether you have further concerns.

sam-mccall reviewed (approved):
Thanks a lot!

BigBahss commented:
Happy to help 👍
